diff --git a/lib/SPIRV/SPIRVWriter.cpp b/lib/SPIRV/SPIRVWriter.cpp
index ec4ec41f..b481609f 100644
--- a/lib/SPIRV/SPIRVWriter.cpp
+++ b/lib/SPIRV/SPIRVWriter.cpp
@@ -401,13 +401,23 @@ SPIRVType *LLVMToSPIRVBase::transType(Type *T) {
   }
 
   if (T->isBFloatTy()) {
-    BM->getErrorLog().checkError(
-        BM->isAllowedToUseExtension(ExtensionID::SPV_KHR_bfloat16),
-        SPIRVEC_RequiresExtension,
-        "SPV_KHR_bfloat16\n"
-        "NOTE: LLVM module contains bfloat type, translation of which "
-        "requires this extension");
-    return mapType(T, BM->addFloatType(16, FPEncodingBFloat16KHR));
+    // Workaround for LTS2 driver.
+    const char *driverVersion = std::getenv("INTEL_XPU_BACKEND_DRIVER_VERSION");
+    if (driverVersion) {
+      int v0 = 0, v1 = 0, v2 = 0, v3 = 0;
+      sscanf(driverVersion, "%d.%d.%d+%d", &v0, &v1, &v2, &v3);
+      std::tuple<int, int, int, int> ver = {v0, v1, v2, v3};
+      std::tuple<int, int, int, int> minVer = {1, 6, 35096, 9};
+      if (ver >= minVer) {
+        BM->getErrorLog().checkError(
+            BM->isAllowedToUseExtension(ExtensionID::SPV_KHR_bfloat16),
+            SPIRVEC_RequiresExtension,
+            "SPV_KHR_bfloat16\n"
+            "NOTE: LLVM module contains bfloat type, translation of which "
+            "requires this extension");
+        return mapType(T, BM->addFloatType(16, FPEncodingBFloat16KHR));
+      }
+    }
   }
 
   if (T->isFloatingPointTy())
