diff --git a/lib/SPIRV/OCLToSPIRV.cpp b/lib/SPIRV/OCLToSPIRV.cpp
index 05a60ec4..2866141c 100644
--- a/lib/SPIRV/OCLToSPIRV.cpp
+++ b/lib/SPIRV/OCLToSPIRV.cpp
@@ -152,8 +152,8 @@ getAtomicPointerMemorySemanticsMemoryMask(const Value *Ptr,
 
 static size_t getOCLCpp11AtomicMaxNumOps(StringRef Name) {
   return StringSwitch<size_t>(Name)
-      .Cases({"load", "flag_test_and_set", "flag_clear"}, 3)
-      .Cases({"store", "exchange"}, 4)
+      .Cases("load", "flag_test_and_set", "flag_clear", 3)
+      .Cases("store", "exchange", 4)
       .StartsWith("compare_exchange", 6)
       .StartsWith("fetch", 4)
       .Default(0);
diff --git a/lib/SPIRV/SPIRVUtil.cpp b/lib/SPIRV/SPIRVUtil.cpp
index b747e423..14fc6d05 100644
--- a/lib/SPIRV/SPIRVUtil.cpp
+++ b/lib/SPIRV/SPIRVUtil.cpp
@@ -620,11 +620,11 @@ static std::string demangleBuiltinOpenCLTypeName(StringRef MangledStructName) {
 /// floating point type.
 static Type *parsePrimitiveType(LLVMContext &Ctx, StringRef Name) {
   return StringSwitch<Type *>(Name)
-      .Cases({"char", "signed char", "unsigned char"}, Type::getInt8Ty(Ctx))
-      .Cases({"short", "unsigned short"}, Type::getInt16Ty(Ctx))
-      .Cases({"int", "unsigned int"}, Type::getInt32Ty(Ctx))
-      .Cases({"long", "unsigned long"}, Type::getInt64Ty(Ctx))
-      .Cases({"long long", "unsigned long long"}, Type::getInt64Ty(Ctx))
+      .Cases("char", "signed char", "unsigned char", Type::getInt8Ty(Ctx))
+      .Cases("short", "unsigned short", Type::getInt16Ty(Ctx))
+      .Cases("int", "unsigned int", Type::getInt32Ty(Ctx))
+      .Cases("long", "unsigned long", Type::getInt64Ty(Ctx))
+      .Cases("long long", "unsigned long long", Type::getInt64Ty(Ctx))
       .Case("half", Type::getHalfTy(Ctx))
       .Case("float", Type::getFloatTy(Ctx))
       .Case("double", Type::getDoubleTy(Ctx))
diff --git a/lib/SPIRV/SPIRVWriter.cpp b/lib/SPIRV/SPIRVWriter.cpp
index b017e663..3fd8f21b 100644
--- a/lib/SPIRV/SPIRVWriter.cpp
+++ b/lib/SPIRV/SPIRVWriter.cpp
@@ -5397,16 +5397,16 @@ LLVMToSPIRVBase::getFPBuiltinType(IntrinsicInst *II, StringRef &OpName) {
   OpName = Name.split('.').first;
   FPBuiltinType Type =
       StringSwitch<FPBuiltinType>(OpName)
-          .Cases({"fadd", "fsub", "fmul", "fdiv", "frem"},
+          .Cases("fadd", "fsub", "fmul", "fdiv", "frem",
                  FPBuiltinType::REGULAR_MATH)
-          .Cases({"sin", "cos", "tan"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"sinh", "cosh", "tanh"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"asin", "acos", "atan"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"asinh", "acosh", "atanh"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"exp", "exp2", "exp10", "expm1"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"log", "log2", "log10", "log1p"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"sqrt", "rsqrt", "erf", "erfc"}, FPBuiltinType::EXT_1OPS)
-          .Cases({"atan2", "pow", "hypot", "ldexp"}, FPBuiltinType::EXT_2OPS)
+          .Cases("sin", "cos", "tan", FPBuiltinType::EXT_1OPS)
+          .Cases("sinh", "cosh", "tanh", FPBuiltinType::EXT_1OPS)
+          .Cases("asin", "acos", "atan", FPBuiltinType::EXT_1OPS)
+          .Cases("asinh", "acosh", "atanh", FPBuiltinType::EXT_1OPS)
+          .Cases("exp", "exp2", "exp10", "expm1", FPBuiltinType::EXT_1OPS)
+          .Cases("log", "log2", "log10", "log1p", FPBuiltinType::EXT_1OPS)
+          .Cases("sqrt", "rsqrt", "erf", "erfc", FPBuiltinType::EXT_1OPS)
+          .Cases("atan2", "pow", "hypot", "ldexp", FPBuiltinType::EXT_2OPS)
           .Case("sincos", FPBuiltinType::EXT_3OPS)
           .Default(FPBuiltinType::UNKNOWN);
   return Type;
