diff --git a/torch/_inductor/runtime/hints.py b/torch/_inductor/runtime/hints.py
index 276c01f3f42..5c633b7963b 100644
--- a/torch/_inductor/runtime/hints.py
+++ b/torch/_inductor/runtime/hints.py
@@ -48,8 +48,8 @@ if _is_triton_available():
         ):
             # Prepare the arguments for AttrsDescriptor
             kwargs = {
-                "tt.divisibility": divisible_by_16,
-                "tt.equal_to": equal_to_1,
+                "tt.divisibility": tuple([(i,) for i in divisible_by_16]),
+                "tt.equal_to": tuple([(i,) for i in equal_to_1]),
             }
 
             # Instantiate AttrsDescriptor with the prepared arguments
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index af8530e94d0..0d0e35d35eb 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -435,11 +435,20 @@ class CachingAutotuner(KernelInterface):
         else:
             triton_helpers.set_driver_to_gpu()
 
+        def fixup_signature(signature, constants):
+            signature = signature.copy()
+            for constant in constants:
+                # If it's not in the signature already, it's a constexpr
+                # argument that we need to add in
+                if constant not in signature:
+                    signature[constant] = "constexpr"
+            return signature
+
         if ASTSource:
             compile_args = (
                 ASTSource(
                     self.fn,
-                    compile_meta["signature"],
+                    fixup_signature(compile_meta["signature"], compile_meta["constants"]),
                     compile_meta["constants"],
                     compile_meta["configs"][0],
                 ),
