diff --git a/test/inductor/test_triton_kernels.py b/test/inductor/test_triton_kernels.py
index 4d7a85029e3..6cf5e05a6f5 100644
--- a/test/inductor/test_triton_kernels.py
+++ b/test/inductor/test_triton_kernels.py
@@ -1268,9 +1268,9 @@ def forward(self, x_1, output_1):
         if dynamic:
             # when half_n_elements passed to the Triton kernel is
             # dynamic, equal_to_1 specializaiton can't be enforced
-            self.assertTrue(_triton_get_ast_equal_to_str(()) in sources[0])
+            self.assertTrue(_triton_get_ast_equal_to_str([]) in sources[0])
         else:
-            self.assertTrue(_triton_get_ast_equal_to_str((3,)) in sources[0])
+            self.assertTrue(_triton_get_ast_equal_to_str([(3,)]) in sources[0])
         self.assertEqual(compiled_out, eager_out)
 
     @requires_gpu
