diff --git a/test/inductor/test_codegen_triton.py b/test/inductor/test_codegen_triton.py
index 84264bf1b01..eb3fac0f398 100644
--- a/test/inductor/test_codegen_triton.py
+++ b/test/inductor/test_codegen_triton.py
@@ -48,7 +48,7 @@ class TestCodegenTriton(InductorTestCase):
                 return config.divisible_by_16
 
         self.assertEqual(
-            (2,),
+            [(2,)],
             _check_divisibility(
                 triton_utils.config_of(
                     [
@@ -63,7 +63,7 @@ class TestCodegenTriton(InductorTestCase):
         )
 
         self.assertEqual(
-            (0, 2, 4, 5, 6),
+            [(0, 2, 4, 5, 6)],
             _check_divisibility(
                 triton_utils.config_of(
                     [
