diff --git a/torch/_inductor/codegen/xpu/device_op_overrides.py b/torch/_inductor/codegen/xpu/device_op_overrides.py
index 99502ca2dd..5d538ec20c 100644
--- a/torch/_inductor/codegen/xpu/device_op_overrides.py
+++ b/torch/_inductor/codegen/xpu/device_op_overrides.py
@@ -61,7 +61,7 @@ class XPUDeviceOpOverrides(DeviceOpOverrides):
     def cpp_scratch(
         self, idx: int, workspace: TritonScratchWorkspace, prefix: Optional[str] = None
     ) -> Optional[tuple[list[str], str]]:
-        return None
+        return [f"void *global_scratch_{idx} = 0;"], f"global_scratch_{idx}"
 
 
 register_device_op_overrides("xpu", XPUDeviceOpOverrides())
