diff --git a/test/inductor/test_torchinductor.py b/test/inductor/test_torchinductor.py
index 1268fcc..0442012 100644
--- a/test/inductor/test_torchinductor.py
+++ b/test/inductor/test_torchinductor.py
@@ -9723,6 +9723,7 @@ def forward(self, arg0_1: "Sym(s77)", arg1_1: "Sym(s27)", arg2_1: "Sym(s53)", ar
         self.common(fn, (inp, False))
         self.common(fn, (inp, True))
 
+    @skipIfXpu
     def test_topk(self):
         def fn(a):
             return torch.topk(a, 2, -1)
