diff --git a/test/inductor/test_triton_kernels.py b/test/inductor/test_triton_kernels.py
--- a/test/inductor/test_triton_kernels.py
+++ b/test/inductor/test_triton_kernels.py
@@ -49,6 +49,7 @@
     HAS_CUDA_AND_TRITON,
     HAS_GPU,
     HAS_XPU_AND_TRITON,
+    requires_block_ptr,
     TRITON_HAS_CPU,
 )
 from torch.testing._internal.logging_utils import log_settings, logs_to_string
@@ -4288,6 +4289,7 @@
             ["O_ptr"],
         )
 
+    @requires_block_ptr
     @make_mutation_test
     def test_for_loop_arg_2():
         @triton.jit
@@ -4348,6 +4350,7 @@
             ["o_ptr"],
         )
 
+    @requires_block_ptr
     @make_mutation_test
     def test_while_loop():
         @triton.jit
@@ -5019,6 +5022,9 @@
         if kernel.fn.__name__ == "add_kernel_2d_autotuned":
             fn = unittest.skip("Fails with Triton update")(fn)
 
+        if "block_ptr" in kernel.fn.__name__:
+            fn = requires_block_ptr(fn)
+
         setattr(MutationTests, name, fn)
 
 
