diff --git a/test/inductor/test_compile_subprocess.py b/test/inductor/test_compile_subprocess.py
index 6eba88ecae9..2e9ab9c06df 100644
--- a/test/inductor/test_compile_subprocess.py
+++ b/test/inductor/test_compile_subprocess.py
@@ -52,9 +52,6 @@ test_failures = {
     "test_remove_noop_slice_scatter": TestFailure(("xpu"), is_skip=True),
     "test_remove_noop_view_default": TestFailure(("xpu"), is_skip=True),
     "test_remove_noop_view_dtype": TestFailure(("xpu"), is_skip=True),
-    # TODO:remove test_upsample_bicubic2d after the following issue resolved:
-    # https://github.com/intel/intel-xpu-backend-for-triton/issues/4184
-    "test_upsample_bicubic2d": TestFailure(("xpu"), is_skip=False),
 }
 
 
